torch/_expanded_weights/expanded_weights_impl.py

```diff
@@ -0,0 +1,203 @@
+
+    @property
+    def dtype(self):
+        return self.orig_weight.dtype
+
+    @property
+    def grad(self):
+        return None
+
+    @grad.setter
+    def grad(self, value):
+        if value is None:
+            return
+        else:
+            raise RuntimeError("ExpandedWeights should never have a grad value set on it.")
+
+    @property
+    def requires_grad(self):
+        return self.orig_weight.requires_grad
+
+    @property
+    def grad_fn(self):
+        return None
+
+    def __hash__(self):
+        return id(self)
+
+    def __repr__(self):
+        return "ExpandedWeight for:\n" + self.orig_weight.__repr__() + f" with batch size {self.batch_size}"
+
+# Fallback:
+# - forward pass: run operation with de-expanded input
+# - backward pass: run autograd.grad in a for-loop to compute per-sample-grads.
+#                  This is NOT something the vmap API can handle, something more
+#                  low-level is at work here.
+class SlowFallback(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, *func_and_expanded_args):
+        func = func_and_expanded_args[0]
+        expanded_args = func_and_expanded_args[1:]
+        with torch.enable_grad():
+            output = forward_helper(func, ctx, expanded_args, 1)
+            true_output = ctx.true_outputs
+            if not isinstance(true_output, torch.Tensor):
+                raise RuntimeError(f"Fallback only works on Tensor output, got output was {type(true_output).__name__}")
+            outputs = tuple(true_output[i] for i in range(true_output.shape[0]))
+        ctx.outputs = outputs
+        return output
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        outputs = ctx.outputs
+        diff_args = tuple(arg for arg in ctx.args
+                          if isinstance(arg, torch.Tensor) and arg.requires_grad)
+        diff_args = tuple(arg.orig_weight if isinstance(arg, ExpandedWeight) else arg for arg in diff_args)
+        batch_size = grad_output.shape[0]
+        per_sample_grads = tuple(torch.autograd.grad(outputs[i], diff_args, grad_output[i],
+                                                     retain_graph=(i != batch_size - 1))
+                                 for i in range(batch_size))
+        per_sample_grads_iter = zip(*per_sample_grads)
+        result: List[Union[torch.Tensor, None]] = []
+        result.append(None)  # for function input
+        for arg in ctx.args:
+            if isinstance(arg, ExpandedWeight):
+                cur_per_sample_grad = next(per_sample_grads_iter)
+                arg.orig_weight.grad_sample = torch.stack(cur_per_sample_grad)
+                result.append(None)
+            elif isinstance(arg, torch.Tensor) and arg.requires_grad:
+                cur_per_sample_grad = next(per_sample_grads_iter)
+                result.append(torch.stack(cur_per_sample_grad).sum(0))
+            else:
+                result.append(None)
+        result.append(None)
```

Comment:
I'm a bit confused about all these Nones, but maybe a comment with an example will help clear that up